test/jdk/java/lang/Thread/crac/ThreadSleepOnCRPauseTest.java

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019-2021 Azul Systems, Inc.  All Rights Reserved.
+// DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+//
+// This code is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License version 2 only, as published by
+// the Free Software Foundation.
+//
+// This code is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+// A PARTICULAR PURPOSE.  See the GNU General Public License version 2 for more
+// details (a copy is included in the LICENSE file that accompanied this code).
+//
+// You should have received a copy of the GNU General Public License version 2
+// along with this work; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+//
+// Please contact Azul Systems, 385 Moffett Park Drive, Suite 115, Sunnyvale,
+// CA 94089 USA or visit www.azul.com if you need additional information or
+// have any questions.
+
+/*
+ * @test ThreadSleepOnCRPauseTest.java
+ * @requires (os.family == "linux")
+ * @library /test/lib
+ * @summary check if the Thread.sleep() will be completed on restore immediately
+ *          if its end time fell on the CRaC pause period
+ *          (i.e. between the checkpoint and restore)
+ * @run main ThreadSleepOnCRPauseTest
+ */
+
+
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+import java.util.concurrent.CountDownLatch;
+
+
+public class ThreadSleepOnCRPauseTest {
+
+    private final static long SLEEP   = 1500; // [ms]
+    private final static long CRPAUSE = 2500; // [ms]
+
+    private final static long EPS = 300_000_000; // [ns], i.e. 0.3 s
+
+    private final CountDownLatch sleepLatch = new CountDownLatch(1);
+
+    private volatile long end1Time = 0;
+    private volatile long end2Time = 0;
+
+
+    private void runTest() throws Exception {
+
+        Runnable r1 = () -> {
+
+            try {
+                sleepLatch.await();
+                Thread.sleep(SLEEP);
+                end1Time = System.nanoTime();
+            } catch (InterruptedException ie) {
+                throw new RuntimeException(ie);
+            }
+        };
+
+        Runnable r2 = () -> {
+
+            try {
+                sleepLatch.await();
+                Thread.sleep(SLEEP, 20);
+                end2Time = System.nanoTime();
+            } catch (InterruptedException ie) {
+                throw new RuntimeException(ie);
+            }
+        };
+
+        Thread t1 = new Thread(r1), t2 = new Thread(r2);
+        t1.start();
+        t2.start();
+        sleepLatch.countDown(); // sleep simultaneously
+
+        while ((t1.getState() != Thread.State.TIMED_WAITING) &&
```

Comment:
we cannot use explicit sync from the sleep method itself.

so, without checking the thread status, it is difficult to avoid the race between sleeps in t1, t2 and checkpoint in main, as the threads could nofify the main only that they are GOING to sleep, not sleeping already

Reply:
Yes, assume some sync before going to sleep. Your argument is valid, but we need to ensure the thread is doing exactly our Thread.sleep, not executing an arbitrary method from the https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Thread.State.html#TIMED_WAITING.

Interesting, we may synchronize threads correctly for Object.wait

```
Object lock = new Object();
bool started = false;

new Thread ({
  synchronized (lock) {
    lock.notify();
    lock.wait(...)
  }
  // after restore
})

synchronized (lock) {
  while (!started) {
    lock.wait();
  }
}
// the other thread is in lock.wait
checkpointRestore();
```

Reply:
> we need to ensure the thread is doing exactly our Thread.sleep not executing an arbitrary method from the ...
yes, but we did not call no one of these methods, but sleep. so i'm not sure if there can bee any mess in principle

wait-notify here has the same effect as the latch, but i can change the latch "direction" to trigger the CR after the "ready-to-sleep" points